src/data/sum.lean

```diff
@@ -111,12 +112,57 @@ funext $ λ x, sum.cases_on x (λ _, rfl) (λ _, rfl)
   sum.elim (f ∘ inl) (f ∘ inr) = f :=
 funext $ λ x, sum.cases_on x (λ _, rfl) (λ _, rfl)
 
-lemma elim_injective {α β γ : Sort*} {f : α → γ} {g : β → γ}
-  (hf : function.injective f) (hg : function.injective g)
- (hfg : ∀ a b, f a ≠ g b) : function.injective (sum.elim f g) :=
-λ x y, sum.rec_on x
-  (sum.rec_on y (λ x y hxy, by rw hf hxy) (λ x y hxy, false.elim $ hfg _ _ hxy))
-  (sum.rec_on y (λ x y hxy, false.elim $ hfg x y hxy.symm) (λ x y hxy, by rw hg hxy))
+open function (update update_eq_iff update_comp_eq_of_injective update_comp_eq_of_forall_ne)
+
+@[simp] lemma update_elim_inl {α β γ} [decidable_eq α] [decidable_eq (α ⊕ β)]
+  {f : α → γ} {g : β → γ} {i : α} {x : γ} :
+  update (sum.elim f g) (inl i) x = sum.elim (update f i x) g :=
+update_eq_iff.2 ⟨by simp, by simp { contextual := tt }⟩
+
+@[simp] lemma update_elim_inr {α β γ} [decidable_eq β] [decidable_eq (α ⊕ β)]
+  {f : α → γ} {g : β → γ} {i : β} {x : γ} :
+  update (sum.elim f g) (inr i) x = sum.elim f (update g i x) :=
+update_eq_iff.2 ⟨by simp, by simp { contextual := tt }⟩
+
+@[simp] lemma update_inl_comp_inl {α β γ} [decidable_eq α] [decidable_eq (α ⊕ β)]
+  {f : α ⊕ β → γ} {i : α} {x : γ} :
+  update f (inl i) x ∘ inl = update (f ∘ inl) i x :=
+update_comp_eq_of_injective _ injective_inl _ _
+
+@[simp] lemma update_inl_apply_inl {α β γ} [decidable_eq α] [decidable_eq (α ⊕ β)]
+  {f : α ⊕ β → γ} {i j : α} {x : γ} :
+  update f (inl i) x (inl j) = update (f ∘ inl) i x j :=
+by rw ← update_inl_comp_inl
+
+@[simp] lemma update_inl_comp_inr {α β γ} [decidable_eq α] [decidable_eq (α ⊕ β)]
+  {f : α ⊕ β → γ} {i : α} {x : γ} :
+  update f (inl i) x ∘ inr = f ∘ inr :=
+update_comp_eq_of_forall_ne _ _ $ λ _, inr_ne_inl
+
+@[simp] lemma update_inl_apply_inr {α β γ} [decidable_eq α] [decidable_eq (α ⊕ β)]
+  {f : α ⊕ β → γ} {i : α} {j : β} {x : γ} :
```

Comment:
Is it worth having these lemmas for when f is a dependent function and gamma is a pi type?

Reply:
There is more than one way to formulate it (e.g., `γ : Π i : α ⊕ β, Sort*` with `γ (sum.inl i)` or `(γ₁ : Π i : α, Sort*) (γ₂ : Π i : β, Sort*)` together with `x : sum.elim γ₁ γ₂ i`). I'd postpone this decision till someone needs the dependent version.

Reply:
Isn't the latter just a special case of the former?

Reply:
I'm not sure that Lean will unfold `sum.elim` in types.

Reply:
I think this is hard to discuss without a full theorem statement. Either way, I'm not objecting to it being postponed. Perhaps worth a TODO comment, but not a big deal.